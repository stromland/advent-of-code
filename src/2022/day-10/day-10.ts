type Operation = 'addx' | 'noop';

type Instruction = {
  operation: Operation;
  value: number;
};

const strengthCycles = [20, 60, 100, 140, 180, 220];

function getSignalStrength(cycles: number, register: number): number {
  return strengthCycles.includes(cycles) ? cycles * register : 0;
}

function parseInstructions(input: string[]): Instruction[] {
  return input.map((it) => {
    const [operation, value] = it.split(' ') as [Operation, string];
    return { operation, value: Number(value) };
  });
}

function runProgram<T>(
  input: string[],
  onInstruction: (x: number) => T,
  onCycle: (cycles: number, x: number, val: T) => number
): void {
  let cycles = 0;
  let x = 1;
  parseInstructions(input).forEach((ins) => {
    const value = onInstruction(x);
    const insCycles = ins.operation === 'addx' ? 2 : 1;
    for (let i = 0; i < insCycles; i++) {
      cycles = onCycle(cycles, x, value);
    }
    if (ins.operation === 'addx') {
      x += ins.value;
    }
  });
}

export function part1(input: string[]): number {
  let strength = 0;
  runProgram(
    input,
    () => 0,
    (cycles, x) => {
      cycles++;
      strength += getSignalStrength(cycles, x);
      return cycles;
    }
  );

  return strength;
}

function getSprite(register: number): string[] {
  return Array.from({ length: 40 }, () => '.').map((it, index) =>
    register - 1 <= index && register + 1 >= index ? '#' : it
  );
}

export function part2(input: string[]): string[][] {
  const display: string[][] = [];
  runProgram(input, getSprite, (cycles, x, sprite) => {
    const pixel = sprite[cycles % 40];
    const row = Math.floor(cycles / 40);
    if (!display[row]) {
      display[row] = [pixel];
    } else {
      display[row].push(pixel);
    }
    return cycles + 1;
  });
  return display;
}
