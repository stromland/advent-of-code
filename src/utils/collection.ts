export function collectUnique<T>(
  arr: Iterable<T> | ArrayLike<T>,
  predicate: (item: T) => boolean = () => true
): Set<T> {
  return Array.from(arr).reduce((set, item) => {
    if (predicate(item)) {
      set.add(item);
    }
    return set;
  }, new Set<T>());
}

export function chunk<T>(
  arr: Iterable<T> | ArrayLike<T>,
  chunkSize: number
): T[][] {
  return Array.from(arr).reduce((list, item) => {
    const index = list.length - 1;

    if (list[index] === undefined || list[index].length === chunkSize) {
      list.push([item]);
      return list;
    }

    list[index].push(item);
    return list;
  }, [] as T[][]);
}

export function sum<T>(
  arr: Iterable<T> | ArrayLike<T>,
  fn: (value: T) => number
): number {
  return Array.from(arr).reduce((total, next) => total + fn(next), 0);
}

export function countTill<T>(arr: T[], predicate: (it: T) => boolean): number {
  for (let i = 0; i < arr.length; i++) {
    if (predicate(arr[i])) {
      return i + 1;
    }
  }
  return arr.length;
}
