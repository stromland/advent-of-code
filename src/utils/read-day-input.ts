import { readFileSync } from 'fs';
import { resolve } from 'path';

type InputType = 'PUZZLE' | 'TEST';

export function readDayInput(
  year: number,
  day: number,
  type: InputType
): string[] {
  const file = getDayInputFile(year, day, type);
  return readFileSync(file).toString().split('\n');
}

function getDayInputFile(year: number, day: number, type: InputType): string {
  const inputType = type === 'PUZZLE' ? 'puzzle-input' : 'test-input';
  return resolve(
    __dirname,
    `../../src/${year}/day-${day}/day-${day}-${inputType}.txt`
  );
}
