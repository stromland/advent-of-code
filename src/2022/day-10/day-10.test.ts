import { readDayInput } from '../../utils/read-day-input';
import { part1, part2 } from './day-10';

const testInput = readDayInput(2022, 10, 'PUZZLE');

describe('day-10', () => {
  test('part1', () => {
    const answer = part1(testInput);
    expect(answer).toBe(10760);
  });

  test('part2', () => {
    const answer = part2(testInput);
    const text = answer.map((it) => it.join('')).join('\n');
    expect(text).toMatchSnapshot();
  });
});
