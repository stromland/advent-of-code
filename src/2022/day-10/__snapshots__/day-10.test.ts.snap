// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`day-10 part2 1`] = `
"####.###...##..###..#..#.####..##..#..#.
#....#..#.#..#.#..#.#..#.#....#..#.#..#.
###..#..#.#....#..#.####.###..#....####.
#....###..#.##.###..#..#.#....#.##.#..#.
#....#....#..#.#....#..#.#....#..#.#..#.
#....#.....###.#....#..#.#.....###.#..#.
."
`;
